src/world/action/sneaky_parasol.c

```diff
@@ -1,52 +1,368 @@
+            playerStatus->targetYaw = clamp_angle(cam->currentYaw - temp_f0_3 - 90.0f);
+            break;
+        case 24:
+            if (--playerStatus->framesOnGround == 0) {
+                set_time_freeze_mode(TIME_FREEZE_NORMAL);
+                playerStatus->flags &= ~0x100000;
+                set_action_state(ACTION_STATE_IDLE);
+                enable_player_static_collisions();
+            }
+            break;
+        case 26:
+            func_800E63A4(0);
+            enable_player_static_collisions();
+            break;
+        case 40:
+            if (playerStatus->unk_BC != 0) {
+                set_action_state(ACTION_STATE_IDLE);
+                enable_player_static_collisions();
+            }
+            break;
+        case 50:
+            if (--playerStatus->framesOnGround == 0) {
+                set_action_state(ACTION_STATE_IDLE);
+                enable_player_static_collisions();
+            }
+            break;
     }
 
-    return ret;
-}
-#else
-INCLUDE_ASM(Npc*, "world/action/sneaky_parasol", func_802B6000_E2A6B0, void);
-#endif
+    if (parasolStruct->unk_08 > 0) {
+        if (--parasolStruct->unk_08 == 0xA) {
+            if (playerStatus->spriteFacingAngle >= 90.0f && playerStatus->spriteFacingAngle < 270.0f) {
+                phi_f20 = ((cam->currentYaw - 270.0f) * TAU) / 360.0f;
+                phi_f22 = 46.0f;
+            } else {
+                phi_f20 = ((cam->currentYaw - 90.0f) * TAU) / 360.0f;
+                phi_f22 = 30.0f;
+            }
+            parasolStruct->position.x = playerStatus->position.x + (phi_f22 * sin_rad(phi_f20));
+            parasolStruct->position.z = playerStatus->position.z - (phi_f22 * cos_rad(phi_f20));
+            parasolStruct->position.y = playerStatus->position.y - 20.0f;
+        }
+        if (parasolStruct->unk_08 < 0xB && parasolStruct->unk_08 & 1) {
+            playFX_11(3, parasolStruct->position.x - 8.0f, parasolStruct->position.y + 50.0f, parasolStruct->position.z, 2.0f);
+            temp_f22 = parasolStruct->position.x;
+            tempX = (((cam->currentYaw + playerStatus->spriteFacingAngle) - 90.0f) * TAU) / 360.0f;
 
-INCLUDE_ASM(s32, "world/action/sneaky_parasol", func_802B6120_E2A7D0);
+            phi_f20 = tempX;
+            tempX = temp_f22 + (sin_rad(phi_f20) * 10.0);
+            tempZ = parasolStruct->position.z;
+
+            parasolStruct->position.x = tempX;
+            parasolStruct->position.z = (tempZ - (cos_rad(phi_f20) * 10.0));
+        }
+    } else if (parasolStruct->unk_08 == 0) {
+        parasolStruct->unk_08 = -1;
+        if (parasolStruct->npc != NULL) {
+            playFX_84(playerStatus->unk_0D - 1, parasolStruct->npc->pos.x, parasolStruct->npc->pos.y, parasolStruct->npc->pos.z, 1.0f, parasolStruct->npc->yawCamOffset);
+        }
+    }
+    if (parasolStruct->unk_0C != 0) {
+        if (parasolStruct->unk_0C < 0xB) {
+            if (parasolStruct->unk_0C == 0xA) {
```

Comment:
This looks like a switch, @ethteck is better at seeing these though so he may correct me 🌮

Reply:
I'm honestly not sure. I think it's okay for now to leave it